src/aleph/http/http2.clj

```diff
@@ -0,0 +1,503 @@
+    [manifold.stream :as s])
+  (:import
+    (aleph.http.file HttpFile)
+    (clojure.lang IPending)
+    (io.netty.buffer ByteBuf)
+    (io.netty.channel FileRegion)
+    (io.netty.handler.codec.http2
+      DefaultHttp2DataFrame
+      DefaultHttp2Headers
+      DefaultHttp2HeadersFrame
+      Http2DataChunkedInput
+      Http2Error
+      Http2FrameStream
+      Http2FrameStreamException
+      Http2Headers
+      Http2StreamChannel)
+    (io.netty.handler.ssl SslHandler)
+    (io.netty.handler.stream
+      ChunkedFile
+      ChunkedInput
+      ChunkedNioFile)
+    (io.netty.util.internal StringUtil)
+    (java.io
+      Closeable
+      File
+      RandomAccessFile)
+    (java.nio ByteBuffer)
+    (java.nio.channels FileChannel)
+    (java.nio.file
+      OpenOption
+      Path
+      StandardOpenOption)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private byte-array-class (Class/forName "[B"))
+
+(def ^:const max-allowed-frame-size 16777215) ; 2^24 - 1, see RFC 9113, SETTINGS_MAX_FRAME_SIZE
+(def ^:dynamic *default-chunk-size* 16384) ; same as default SETTINGS_MAX_FRAME_SIZE
+
+;; See https://httpwg.org/specs/rfc9113.html#ConnectionSpecific
+(def invalid-headers #{"connection" "proxy-connection" "keep-alive" "upgrade"})
+
+(let [illegal-arg (fn [^String emsg]
+                    (let [ex (IllegalArgumentException. emsg)]
+                      (log/error ex emsg)
+                      (throw ex)))]
+  (defn- add-header
+    "Add a single header and value. The value can be a string or a collection of
+     strings.
+
+     Respects HTTP/2 rules. Strips invalid connection-related headers. Throws on
+     nil header values. Throws if `transfer-encoding` is present, but is not
+     equal to 'trailers'."
+    [^Http2Headers h2-headers ^String header-name header-value]
+    (log/debug "Adding HTTP header" header-name ": " header-value)
+
+    (if (nil? header-name)
+      (throw (IllegalArgumentException. "Header name cannot be nil"))
+      (let [header-name (str/lower-case header-name)]       ; http2 requires lowercase headers
+        (cond
+          (nil? header-value)
+          (illegal-arg (str "Invalid nil value for header '" header-name "'"))
+
+          (invalid-headers header-name)
+          (illegal-arg (str "Forbidden HTTP/2 header: \"" header-name "\""))
+
+          ;; See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
+          (and (.equals "transfer-encoding" header-name)
+               (not (.equals "trailers" header-value)))
+          (illegal-arg "Invalid value for 'transfer-encoding' header. Only 'trailers' is allowed.")
+
+          (sequential? header-value)
+          (.add h2-headers ^CharSequence header-name ^Iterable header-value)
+
+          :else
+          (.add h2-headers ^CharSequence header-name ^Object header-value))))))
+
+(defn- ring-map->netty-http2-headers
+  "Builds a Netty Http2Headers object from a Ring map."
+  ^DefaultHttp2Headers
+  [req]
```

Comment:
Zach felt that these were hot-enough paths in the HTTP1 code to warrant using `get` directly, and use a ConcurrentHashMap to cache header case transformations, so I did the same for HTTP2. 

I haven't profiled (which I should), but it makes sense. This function will be called for every single request, so especially on the server, it will be hot, and destructuring is (relatively) slow.

If anything, I'm incline to do it more once we profile.

Reply:
Added use of ConcurrentHashMap to HTTP2 headers. Profiling suggests retrieving already-lower-cased headers is 1/3 faster (~750 ns vs 1200 ns) than always lower-casing.